cocos/base/Scheduler.cpp

```diff
@@ -177,39 +151,36 @@ void Scheduler::schedule(const ccSchedulerFunc &callback, void *target, float in
     CCASSERT(target, "Argument target must be non-nullptr");
     CCASSERT(!key.empty(), "key should not be empty!");
 
-    tHashTimerEntry *element = nullptr;
-    HASH_FIND_PTR(_hashForTimers, &target, element);
-
-    if (!element) {
-        element = (tHashTimerEntry *)calloc(sizeof(*element), 1);
+    HashTimerEntry *element = nullptr;
+    auto            iter    = _hashForTimers.find(&target);
+    if (iter == _hashForTimers.end()) {
+        element         = new HashTimerEntry();
         element->target = target;
 
-        HASH_ADD_PTR(_hashForTimers, target, element);
+        _hashForTimers[target] = element;
 
         // Is this the 1st element ? Then set the pause level to all the selectors of this target
         element->paused = paused;
     } else {
         CCASSERT(element->paused == paused, "element's paused should be paused!");
     }
 
-    if (element->timers == nullptr) {
-        element->timers = ccArrayNew(10);
+    if (element->timers.empty()) {
+        element->timers.reserve(INITIAL_TIMER_COUND);
     } else {
-        for (int i = 0; i < element->timers->num; ++i) {
-            TimerTargetCallback *timer = dynamic_cast<TimerTargetCallback *>(element->timers->arr[i]);
-
-            if (timer && key == timer->getKey()) {
+        for (auto &e : element->timers) {
+            auto *timer = dynamic_cast<TimerTargetCallback *>(e);
+            if (key == timer->getKey()) {
                 CC_LOG_DEBUG("CCScheduler#scheduleSelector. Selector already scheduled. Updating interval from: %.4f to %.4f", timer->getInterval(), interval);
                 timer->setInterval(interval);
                 return;
             }
         }
-        ccArrayEnsureExtraCapacity(element->timers, 1);
     }
 
-    TimerTargetCallback *timer = new (std::nothrow) TimerTargetCallback();
+    auto *timer = new (std::nothrow) TimerTargetCallback();
     timer->initWithCallback(this, callback, target, key, interval, repeat, delay);
-    ccArrayAppendObject(element->timers, timer);
+    element->timers.push_back(timer);
     timer->release();
```

Comment:
`timer` should not be released here, since it can nolonger be retained in `ccArrayAppendObject`

Reply:
Yep, you are right.